### ISSUE TITLE
Clarify path to published tables on OEP

### ISSUE THREAD

**Issue by nesnoj** [openego/eGon-data#457]
Clarify the location where tables will be published on the OEP #367. This is required to set the `id` variable in the metadata. (As we use the OEP's current naming in our DB, I assume we can simply keep the schemas' and tables' names.)

It currently follows the pattern `https://openenergy-platform.org/dataedit/view/SCHEMA/TABLE`, this will be kept in the future too, right @christian-rli ?

(You can find a full list of current tables in #450)

**Comment by christian-rli** [comment]
We (@MGlauer @wingechr ) are currently working on dissolving schemas on the OEP. Reasons: schemas are inflexible and annoying. It's hard to move between them. tables cannot belong to more than one and they are hard to find. The goal is to allow users to select their own data collection(s) to which a dataset can belong. All existing schemas will remain as data collections and all existing tables will remain available under the old URL, but new URLs will no longer have the `/SCHEMA` bit in the URL. The table name will be replaced by a number, however we were thinking to allow specific alternative (human readable) URLs to be used as alternative. I'm assuming you would like to make use of that option? 

It's unclear, how long it will take to implement these changes, but we're slowly starting with the process now. I would strongly assume, that if you publish your data this year, they will still fall under the current system.

**Comment by nesnoj** [comment]
The idea of data collections apart from simple tags sounds great, I guess that'll be quiet some work for you (or @Ludee ).. :hammer: 
I think we will be ready to publish the data in spring 2022, we can use whatever option will be available by then.

**Comment by jh-RLI** [comment]
The schema dissolve is still not finished and is now picked up again (and finally finically backed) so will be done until the end of the year. I will implement the dataset API update and i get support with the restructuring of the databse schemas. To make this clear - this is more of a database update, users should not be effected too much aside from the new feature to add table resources to multiple topics (new name for schemas).

-> General issue used for Database schema dissolve and related updates: https://github.com/OpenEnergyPlatform/oeplatform/issues/1356

The api usage will be updated but the old api-pattern will be redirected to the new one. In any case what we do now will also work after the update.

Here is a bit of a [collection where we want to go](https://openenergyplatform.github.io/oeplatform/oeplatform-code/features/oep-datasets/datasets/#implementation), what will change. Mainly, another API endpoint will be added which can be used to create datasets and connect multiple table resources to the dataset. The dataset/resource (table) endpoint will be changed a bit and only needs the table name. When publishing the table from the default to a schema aka topic the name is required to assign the table to a topic.

**Then regarding the general publishing process on the OEP**:

Data is first always grouped into the "model_draft" topic (even after the update). There it will stay until the upload is finalized, metadata is added and especially a proper open data license was added (we use SPDX license list to check that). This (open license) is a hard requirement, once the check succeeds we can publish the data to any of the available topics. Which will also create a PID for the dataset on https://databus.openenergyplatform.org/

**Comment by jh-RLI** [comment]
One more commend, more related to the initial question:
- Yes we can use the schema names, even after the OEP update (which did not happen yet)
- Yes the OEP requires an ID field, will add one if missing

All my previous commends are missing one part. The egon-data produces quite a large volume of data -> at least when working with data in the context of the internet and its limited transmitting capabilities. Meaning the Upload is one crucial part if the full dataset should be available.

.... sorry, this will be a longer read with technical details included. Some of this might become an separate issue later on.

## How we do it for eGon dataset

In egon we introduced a process for that. It includes manual efforts when transporting the data to the Hosting location to avoid transmitting it via the internet. 

Technically, once the data is available locally, the Process is like this:

1 Get access to the TOEP/TOEDB server
2 Create a upload script for your data which uses the DB API to insert data into the TOEDB
3 Run a Django management command to register manually inserted data tables in the Django app (oeplatform) `python manage.py mirror`
4 Run a script to add user permissions to the manually inserted tables
5 Create and Upload metadata using the TOEP metadata WEB-API
6 Verify everything is complete, **especially if a open data license was added to each table resource**
7 Create a SQL-Dump & add everything to the OEDB
8 Repeat steps 3 -6 on OEP 
9 Publish all table resources into the topic (schema) it belongs to using the WEB-API of the OEP.

## What we should do / Test out to enhance the path

This is not perfect but for larger data this might be the only feasible solution for now. Yet it was not fully verified that other solutions will not work. in reGon i would like to test another path which would roughly work like this:

- Avoid backend access to the DB API directly to upload data
- Use the OEP WEB-API (HTTP requests) to:
  - Create a table
  - Upload in chunks of 5000 rows per Request
  - Continuo upload in case of failure
- Start the upload as soon as possible
  - It might be possible to start uploading a table resource once the pipeline run is finished with that specific resource
  - Start it once the pipeline is finished
- Creating metadata, adding an open data license and using the WEB-API to upload the metadata to each table resource is already nicely doable with automation so these steps will stay the same
  - Automating the metadata creation is not fully doable


## OEP Upload case study

I did a little case study to check whether the WEB-API can be used for such a use case. Here are some stats:

- Upload of 1 Mio. rows of data 
  - Only 5 columns (with id col), quite simple data
  - Different data types
  - Always the same data per row
  - Result: https://openenergyplatform.org/dataedit/view/model_draft/chunk_upload_unitl_fail_test
- Took ~1h:10min with my internet connection from home (in theory 30-40 Mbps upload) and from WSL which might further limit the uplink (would have to look into this)
- Using chunks of 5000 rows per Request, sending request with pythons requests package
  - ~ 200 requests until 1 mio goal reached
  - No failure / request limit so far
- No reading from database involved, only generating the chunks in code
- IF we assume that we can upscale this approach, a larger dataset with 1TB will have something around 300 to 400 Mio. rows for "typical" data for our research work, it would take around **13 - 20 days to upload** using this approach
  -  Size of data heavily depends on the data itself 
  - Upload time is reasonable given the size and time it takes to generate the data
  - Maybe it is possible to generate and upload in parallel


This is the script i used

``` python
import requests

LOCAL_OEP_URL = "http://127.0.0.1:8000"
PROD_OEP_URL = "https://openenergyplatform.org/"

HEADER = {"Authorization": "Token <OEP-API-TOKEN>"}

ADVANCED_API = "/api/v0/advanced/search"

REST_API = "/api/v0"

schema = "model_draft"
table = "chunk_upload_unitl_fail_test"


CREATE_NEW_ROW = f"/schema/{schema}/tables/{table}/rows/new"

# response = requests.post(
#     f"{LOCAL_OEP_URL}{ADVANCED_API}"
# )
table_schema = {
    "columns": [
        # NOTE: first column should be numerical column named `id` .
        # Use `bigserial` if you want the database to create the re
        {"name": "id", "data_type": "bigint", "primary_key": True},
        {"name": "name", "data_type": "varchar(18)", "is_nullable": False},
        {"name": "is_active", "data_type": "boolean"},
        {"name": "capacity_mw", "data_type": "float"},
        {"name": "installation_datetime_utc", "data_type": "datetime"},
    ]
}

table_data = {
    # NOTE: first column should be numerical column named `id` .
    # Use `bigserial` if you want the database to create the re
    # {"name": "id", "data_type": "bigserial", "primary_key": True},
    "name": "test",
    "is_active": "true",
    "capacity_mw": "12.12",
    "installation_datetime_utc": "2023-12-12",
    # "json_col": {"key": "value", "list": ["value2", "2"], "list_int": [1, 2, 3]},
}

data_chunk = [table_data.copy() for _ in range(5000)]

url = f"{PROD_OEP_URL}{REST_API}{CREATE_NEW_ROW}"
print(url)

for i in range(10000):
    print(i)
    response = requests.post(
        url=f"{PROD_OEP_URL}{REST_API}{CREATE_NEW_ROW}",
        json={"query": data_chunk},
        headers=HEADER,
    )

    print(response.status_code)
    print(response.json())
    print(response.reason)

```
